walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,187 @@
+    List<TimeRange> availableOptionalTimeRanges = getAvailableTimeRanges(
+        unavailableOptionalTimeRanges, meetingDuration);
+    
+    if (request.getAttendees().isEmpty()) {
+      return availableOptionalTimeRanges;
+    } else if (request.getOptionalAttendees().isEmpty() || availableOptionalTimeRanges.isEmpty()) {
+      return availableTimeRanges;
+    } else {
+      return mergeAvailableTimeRanges(availableTimeRanges, availableOptionalTimeRanges, meetingDuration);
+    }
+}
+
+  // Return a List of TimeRanges containing all TimeRanges that are
+  // unavailable in as little TimeRanges as possible.
+  private List<TimeRange> getUnavailableTimeRanges(Collection<Event> events, Collection<String> attendees) {
+    List<TimeRange> unavailableTimeRanges = new ArrayList<>();
+
+    // If an event's attendee list contains a names of an attendee that needs to attend
+    // (from Collection<String> attendees, then we add the event's TimeRange to a list.
+    events.forEach(event -> {
+      Collection<String> eventAttendants = event.getAttendees();
+
+      for (String attendee : eventAttendants) {
+        if (attendees.contains(attendee)) {
+          TimeRange currentEventTimeRange = event.getWhen();
+          
+          // If the currentEventTimeRange isn't merged into a TimeRange within the
+          // unavailableTimeRanges list, then add the currentEventTimeRange to the list.
+          if (!mergeOnOverlap(unavailableTimeRanges, currentEventTimeRange)) {
+            unavailableTimeRanges.add(currentEventTimeRange);
+          }
+
+          break;
+        }
+      }
+    });
+
+    return unavailableTimeRanges;
+  }
+
+  private boolean mergeOnOverlap(Collection<TimeRange>timeRanges, TimeRange newTimeRange) {
+    // The default value is set to false, which is converted to true if a merge takes place.
+    boolean overlaps = false;
+
+    // Store items that we will add and remove from the collection after we finish iterating
+    // through it in order to avoid ConcurrentModificationException from being thrown
+    List<TimeRange> toRemove = new ArrayList<>();
+    List<TimeRange> toAdd = new ArrayList<>();
+
+    for (TimeRange takenTimeRange : timeRanges) {
+      if (newTimeRange.overlaps(takenTimeRange)) {
+        TimeRange combinedTimeRange = mergeTimeRanges(newTimeRange, takenTimeRange);
+        toRemove.add(takenTimeRange);
+        toAdd.add(combinedTimeRange);
+        overlaps = true;
+      }
+    }
+
+    for (TimeRange timeRange : toRemove) {
+      timeRanges.remove(timeRange);
+    }
+
+    for (TimeRange timeRange : toAdd) {
+      timeRanges.add(timeRange);
+    }
+
+    return overlaps;
+  }
+
+  private TimeRange mergeTimeRanges(TimeRange firstTimeRange, TimeRange secondTimeRange) {
+    TimeRange mergedTimeRange = TimeRange.fromStartEnd(
+        Math.min(firstTimeRange.start(), secondTimeRange.start()),
+        Math.max(firstTimeRange.end(), secondTimeRange.end()),
+        false
+    );
```

Comment:
This line can go to the previous line. Like false);

Reply:
Done.